### PR TITLE
Eng 11879

### PR DESCRIPTION
**Pull request by GuaGuaDean**

### CODE REVIEW

File: src/frontend/org/voltcore/agreement/AgreementSite.java [hunk 1]
```diff
@@ -353,7 +356,7 @@ private void sendHeartbeats(Set<Long> hsIds) {
     private long m_lastHeartbeatTime = System.nanoTime();
     private void processMessage(VoltMessage message) throws Exception {
         if (!m_hsIds.contains(message.m_sourceHSId)) {
-            m_recoveryLog.info("Dropping message " + message + " because it is not from a known up site");
+            logger.log("Dropping message " + message + " because it is not from a known up site",System.currentTimeMillis(), Level.INFO);
```
Comment (luyangco) [inline src/frontend/org/voltcore/agreement/AgreementSite.java]: Please make it a separate pull request, as required by ENG-12169.

Comment (luyangco) [inline src/frontend/org/voltcore/agreement/AgreementSite.java]: Another thing is you can use `RateLimitedLogger.tryLogForMessage()` API to wrap the logger.

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 2]
```diff
@@ -266,6 +267,7 @@
     private ScheduledFuture<?> resMonitorWork;
     private HealthMonitor m_healthMonitor;
 
+    private FailedLoginCounter[] m_flcArray = new FailedLoginCounter[128];
```
Comment (GuaGuaDean) [inline src/frontend/org/voltdb/RealVoltDB.java]: I am thinking about in Line 270 I should avoid using 128, a hardcoded value, as the size of array. Maybe to use a static final int value?

File: src/frontend/org/voltdb/utils/FailedLoginCounter.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+    public int getCount(String user) {
+        if (userFailedAttempts.containsKey(user)) {
+            return userFailedAttempts.get(user);
+        }
+        return 0;
+    }
+
+    public void logMessage(long timestamp, String user) {
+        System.out.println(Thread.currentThread().getId());
```
Comment (luyangco) [inline src/frontend/org/voltdb/utils/FailedLoginCounter.java]: Remove `System.out.println()` when you finish debugging.

File: src/frontend/org/voltdb/utils/FailedLoginCounter.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+import java.text.ParseException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+
+public class FailedLoginCounter {
+    // key is username, value is number of attempts
+    private Map<String,Integer> userFailedAttempts;
+    // key is time bucket, value is user attempts
```
Comment (luyangco) [inline src/frontend/org/voltdb/utils/FailedLoginCounter.java]: Please update your comment to better reflect the data structure. Value is username and failed logging attempts count pair.

File: src/frontend/org/voltdb/utils/FailedLoginCounter.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    public int getCount(String user) {
+        if (userFailedAttempts.containsKey(user)) {
+            return userFailedAttempts.get(user);
+        }
+        return 0;
+    }
+
+    public void logMessage(long timestamp, String user) {
+        System.out.println(Thread.currentThread().getId());
+        if (buckets.containsKey(timestamp)) {
```
Comment (luyangco) [inline src/frontend/org/voltdb/utils/FailedLoginCounter.java]: The millisecond granularity is too fine-grained. Think about it, if there is a continuous DDoS attack going on, and for every second you'll create 1000 new entries in the `bucket`, if your aging out window is 1 minute, then in total 60,000 entries will be in memory before you get a chance to remove some. Why not choose second as the key since your window is one minute.

File: src/frontend/org/voltdb/utils/FailedLoginCounter.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            Map<String,Integer> bucket = buckets.get(timestamp);
+            int bucketCount = bucket.getOrDefault(user,0) + 1;
+              bucket.put(user,bucketCount);
+        } else {
+            buckets.put(timestamp, new HashMap<String,Integer>());
+            buckets.get(timestamp).put(user, 1);
+        }
+        int totalCount = userFailedAttempts.getOrDefault(user,0) + 1;
+        userFailedAttempts.put(user,totalCount);
```
Comment (luyangco) [inline src/frontend/org/voltdb/utils/FailedLoginCounter.java]: In here, you can print something likes `"user XX failed to authenticate XX times in last minute"`. Also remember to rate-limit the print to, let's say, 10 seconds.

File: src/frontend/org/voltdb/utils/FailedLoginCounter.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    public void checkCounter(Long timestamp) throws ParseException {
```
Comment (luyangco) [inline src/frontend/org/voltdb/utils/FailedLoginCounter.java]: I don't see `checkCounter()` be called anywhere, how do you delete old entries when they passed timing window?
One way of doing that, is to submit a periodical task to the executor service, in the task you can call `checkCounter()` to scan then delete old entries.

File: src/frontend/org/voltdb/utils/FailedLoginCounter.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+import java.util.Map;
+import java.util.Map.Entry;
+
+public class FailedLoginCounter {
+    // key is username, value is number of attempts
+    private Map<String,Integer> userFailedAttempts;
+    // key is time bucket, value is user attempts
+    private Map<Long,Map<String,Integer>> buckets;
+
+    final long ONE_MINUTE_IN_MILLIS = 60000;//millisecs
```
Comment (luyangco) [inline src/frontend/org/voltdb/utils/FailedLoginCounter.java]: It would be nice if this can be a tunable. Also please give it a meaningful name.

File: src/frontend/org/voltdb/utils/FailedLoginCounter.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            m_buckets.get(timestampSeconds).put(user, 1);
+        }
+        int totalCount = m_userFailedAttempts.getOrDefault(user,0) + 1;
+        String messageFormat = "user "+ user +" failed to authenticate " + totalCount + " times in last minute";
+        RateLimitedLogger.tryLogForMessage(timestampMilis,
+                10000,
+                TimeUnit.MILLISECONDS,
+                authLog,
+                Level.WARN,
+                messageFormat);
```
Comment (luyangco) [inline src/frontend/org/voltdb/utils/FailedLoginCounter.java]: Your use of RatelimitedLogger is wrong,  your `messageFormat` changes constantly, ratelimited logger will print the log every time it changes. The right way is to pass the messageFormat and parameter separately.

**Comment by luyangco** [comment]
I left some comments for minor issues, and the bigger challenge is how to count fast and accurate.

File: src/frontend/org/voltdb/utils/FailedLoginCounter.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+import org.voltcore.logging.Level;
+import org.voltcore.logging.VoltLogger;
+import org.voltcore.utils.RateLimitedLogger;
+
+public class FailedLoginCounter {
+    // key is username, value is number of failed logging attempts count
+    private Map<String,Integer> m_userFailedAttempts;
+    // key is time bucket, value is username/failed logging attempts count pair
+    private Map<Integer,Map<String,Integer>> m_buckets;
+
```
Comment (GuaGuaDean) [inline src/frontend/org/voltdb/utils/FailedLoginCounter.java]: Shall I print this totalFailedAttempts in the log as well as ratelimited?

File: src/frontend/org/voltdb/utils/FailedLoginCounter.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+    public void logMessage(long timestampMilis, String user) {
+        checkCounter((int)timestampMilis/1000);
+        int timestampSeconds = (int)(timestampMilis/1000);
+        if (m_timeBucketQueue.peekLast().m_ts == timestampSeconds) {
+            TimeBucket bucket = m_timeBucketQueue.peekLast();
+            int bucketCount = bucket.m_userFailedAttempts.getOrDefault(user,0) + 1;
+            bucket.m_userFailedAttempts.put(user,bucketCount);
+        } else {
+            TimeBucket bucket = new TimeBucket();
```
Comment (luyangco) [inline src/frontend/org/voltdb/utils/FailedLoginCounter.java]: I don't see where you add the new `bucket` to `m_timeBucketQueue`.

File: src/frontend/org/voltdb/utils/FailedLoginCounter.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+                ONE_MINUTE_IN_MILLIS,
+                TimeUnit.MILLISECONDS,
+                authLog,
+                Level.WARN,
+                messageFormat,
+                m_totalFailedAttempts);
+    }
+
+    public void checkCounter(int timestamp) {
+        while (!m_timeBucketQueue.isEmpty() && m_timeBucketQueue.peek().m_ts < timestamp) {
```
Comment (luyangco) [inline src/frontend/org/voltdb/utils/FailedLoginCounter.java]: Should be 
```Java
while (!m_timeBucketQueue.isEmpty() && m_timeBucketQueue.peek().m_ts < (timestamp - ONE_MINUTE_IN_MILLIS))
```

File: src/frontend/org/voltdb/utils/FailedLoginCounter.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.TimeUnit;
+
+import org.voltcore.logging.Level;
+import org.voltcore.logging.VoltLogger;
+import org.voltcore.utils.RateLimitedLogger;
+
+public class FailedLoginCounter {
```
Comment (luyangco) [inline src/frontend/org/voltdb/utils/FailedLoginCounter.java]: Please create a unit test for `FailedLoginCounter`.

File: src/frontend/org/voltdb/utils/FailedLoginCounter.java [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    public class TimeBucket {
+        int m_totalFailedAttempts;
+        int m_ts;
```
Comment (luyangco) [inline src/frontend/org/voltdb/utils/FailedLoginCounter.java]: Change the type of timestamp to `Long`, so you don't need to cast in `checkCounter()` and `logMessage()`

File: src/frontend/org/voltdb/utils/FailedLoginCounter.java [hunk 15]
```diff
@@ -50,34 +54,49 @@ public TimeBucket () {
...
-        m_userFailedAttempts.put(user,userFailedCount);
+        m_userFailedAttempts.put(user, userFailedCount);
+
+        int ipFailedCount = m_ipFailedAttempts.getOrDefault(ip, 0) + 1;
+        messageFormat = "IP address "+ ip +" failed to authenticate %d times in last minute";
+        RateLimitedLogger.tryLogForMessage(timestampMilis,
+                                           10000,
+                                           TimeUnit.MILLISECONDS,
+                                           authLog,
+                                           Level.WARN,
```
Comment (luyangco) [inline src/frontend/org/voltdb/utils/FailedLoginCounter.java]: All of those log messages should be INFO level.
